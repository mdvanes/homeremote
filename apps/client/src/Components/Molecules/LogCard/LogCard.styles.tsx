import { makeStyles, Theme, StyleRules, createStyles } from "@mui/material";

const useStyles = makeStyles(
    ({ typography }: Theme): StyleRules =>
        createStyles({
            root: {
                height: "100vh",
                maxHeight: "335px",
                display: "flex",
                flexDirection: "column",
            },
            content: {
                flex: 1,
                overflowX: "auto",
                "& p": {
                    fontSize: typography.fontSize * 0.8,
                },
            },
            message: {
                "& svg": {
                    transform: "translateY(3px)",
                    marginRight: "0.2rem",
                    height: "14px",
                    width: "14px",
                },
            },
            version: {
                paddingBottom: 0,
                "& p": {
                    fontSize: typography.fontSize * 0.7,
                },
            },
        })
);

export default useStyles;
