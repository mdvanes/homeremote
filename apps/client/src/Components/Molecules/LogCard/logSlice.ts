import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { urlToMusicApi } from "../../../Services/urlToMusicApi";

export enum Severity {
    INFO,
    ERROR,
}

export interface Logline {
    message: string;
    severity: Severity;
}

export interface LogState {
    lines: Logline[];
    urgentMessage: Logline | false;
}

export const initialState: LogState = {
    lines: [],
    urgentMessage: false,
};

const writeLog = (...messages: string[]): string => {
    const time = new Date();
    const timestamp = time.toTimeString().substring(0, 8);
    return `${timestamp} ${messages.join(" ")}`;
};

const logSlice = createSlice({
    name: "loglines",
    initialState,
    reducers: {
        logInfo: (draft, { payload }: PayloadAction<string>): void => {
            draft.lines.push({
                message: writeLog("INFO: ", payload),
                severity: Severity.INFO,
            });
        },
        logUrgentInfo: (draft, { payload }: PayloadAction<string>): void => {
            if (payload) {
                const message = writeLog("INFO:", payload);
                const logline = { message, severity: Severity.INFO };
                draft.lines.push(logline);
                draft.urgentMessage = logline;
            }
        },
        logError: (draft, { payload }): void => {
            if (payload) {
                const message = writeLog("ERROR:", payload);
                const logline = { message, severity: Severity.ERROR };
                draft.lines.push(logline);
                draft.urgentMessage = logline;
            }
        },
        clearLog: (): LogState => {
            return initialState;
        },
    },
    extraReducers: (builder) => {
        builder.addMatcher(
            urlToMusicApi.endpoints.getInfo.matchFulfilled,
            (draft, { payload }): void => {
                draft.lines.push({
                    message: writeLog(
                        `INFO: music bin v${payload.versionInfo}`
                    ),
                    severity: Severity.INFO,
                });
            }
        );
        builder.addMatcher(
            urlToMusicApi.endpoints.getMusicProgress.matchFulfilled,
            (draft, { payload }): void => {
                if (payload.state === "finished") {
                    draft.lines.push({
                        message: writeLog(`INFO: Music in v${payload.path}`),
                        severity: Severity.INFO,
                    });
                }
            }
        );
    },
});

export const { logInfo, logUrgentInfo, logError, clearLog } = logSlice.actions;

export default logSlice.reducer;
