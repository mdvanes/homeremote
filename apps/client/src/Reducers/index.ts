import loglinesReducer from "../Components/Molecules/LogCard/logSlice";
import appStatusReducer from "../Components/Molecules/AppStatusButton/appStatusSlice";
import switchBarListReducer from "../Components/Molecules/SwitchBarList/switchBarListSlice";
import authenticationReducer from "../Components/Providers/Authentication/authenticationSlice";
import urlToMusicReducer from "../Components/Molecules/UrlToMusic/urlToMusicSlice";
import { dataloraApi } from "../Services/dataloraApi";
import { dockerListApi } from "../Services/dockerListApi";
import { downloadListApi } from "../Services/downloadListApi";
import { activeConnectionsApi } from "../Services/activeConnectionsApi";
import { combineReducers } from "redux";
import { serviceLinksApi } from "../Services/serviceLinksApi";
import { jukeboxApi } from "../Services/jukeboxApi";
import { urlToMusicApi } from "../Services/urlToMusicApi";
import { monitApi } from "../Services/monitApi";

const rootReducer = combineReducers({
    loglines: loglinesReducer,
    appStatus: appStatusReducer,
    switchesList: switchBarListReducer,
    authentication: authenticationReducer,
    urlToMusic: urlToMusicReducer,
    [activeConnectionsApi.reducerPath]: activeConnectionsApi.reducer,
    [dataloraApi.reducerPath]: dataloraApi.reducer,
    [dockerListApi.reducerPath]: dockerListApi.reducer,
    [downloadListApi.reducerPath]: downloadListApi.reducer,
    [jukeboxApi.reducerPath]: jukeboxApi.reducer,
    [monitApi.reducerPath]: monitApi.reducer,
    [serviceLinksApi.reducerPath]: serviceLinksApi.reducer,
    [urlToMusicApi.reducerPath]: urlToMusicApi.reducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
