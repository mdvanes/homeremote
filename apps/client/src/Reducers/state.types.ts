interface ApiBaseState {
    isLoading: boolean;
    error: string | false;
}

export default ApiBaseState;
