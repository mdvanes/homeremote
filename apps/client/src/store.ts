import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/query";
import { useDispatch } from "react-redux";
import rootReducer from "./Reducers";
import { activeConnectionsApi } from "./Services/activeConnectionsApi";
import { carTwinApi } from "./Services/carTwinApi";
import { dataloraApi } from "./Services/dataloraApi";
import { dockerListApi } from "./Services/dockerListApi";
import { downloadListApi } from "./Services/downloadListApi";
import { energyUsageApi } from "./Services/energyUsageApi";
import { homesecApi } from "./Services/homesecApi";
import { jukeboxApi } from "./Services/jukeboxApi";
import { monitApi } from "./Services/monitApi";
import { nextupApi } from "./Services/nextupApi";
import { nowplayingApi } from "./Services/nowplayingApi";
import { scheduleApi } from "./Services/scheduleApi";
import { serviceLinksApi } from "./Services/serviceLinksApi";
import { stacksApi } from "./Services/stacksApi";
import { urlToMusicApi } from "./Services/urlToMusicApi";

export const store = configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(
            activeConnectionsApi.middleware,
            carTwinApi.middleware,
            dataloraApi.middleware,
            dockerListApi.middleware,
            downloadListApi.middleware,
            energyUsageApi.middleware,
            homesecApi.middleware,
            jukeboxApi.middleware,
            monitApi.middleware,
            nextupApi.middleware,
            nowplayingApi.middleware,
            scheduleApi.middleware,
            serviceLinksApi.middleware,
            stacksApi.middleware,
            urlToMusicApi.middleware
        ),
});

setupListeners(store.dispatch);

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = (): AppDispatch => useDispatch<AppDispatch>();
