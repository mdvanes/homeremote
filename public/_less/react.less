// for /react test

// Toggle
.btn-icon {
    font-size: 350%;
}

// Button-Set
.btn-group-justified {
    > a,
    > span {
        font-size: 230%;
    }
}

.margin-top {
    margin-top: 1em;
}

.log {
    color: #fff;
    height: 300px;
}