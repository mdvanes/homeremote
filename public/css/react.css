.btn-icon {
  font-size: 350%;
}
.btn-group-justified > a,
.btn-group-justified > span {
  font-size: 230%;
}
.margin-top {
  margin-top: 1em;
}
.log {
  color: #fff;
  height: 300px;
}
