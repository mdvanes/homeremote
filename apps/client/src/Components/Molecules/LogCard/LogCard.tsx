import { InfoOutlined, Warning } from "@mui/icons-material";
import {
    Button,
    Card,
    CardActions,
    CardContent,
    Typography,
} from "@mui/material";
import { FC } from "react";
import { useDispatch, useSelector } from "react-redux";
// eslint-disable-next-line @nx/enforce-module-boundaries
import packageJson from "../../../../../../package.json";
import { RootState } from "../../../Reducers";
import gitinfoJson from "../../../gitinfo.json";
import useStyles from "./LogCard.styles";
import { LogState, Severity, clearLog } from "./logSlice";

const SeverityIcons = {
    [Severity.INFO]: (
        <InfoOutlined
            style={{
                fill: "grey",
            }}
        />
    ),
    [Severity.ERROR]: (
        <Warning
            style={{
                fill: "red",
            }}
        />
    ),
};

const Log: FC = () => {
    const { classes } = useStyles();
    const dispatch = useDispatch();
    const loglines = useSelector<RootState, LogState["lines"]>(
        (state: RootState) => state.loglines.lines
    );
    const handleClearLog = (): void => {
        dispatch(clearLog());
    };
    const version =
        gitinfoJson.branch === packageJson.version
            ? `${gitinfoJson.hash}-${packageJson.version}`
            : `${gitinfoJson.hash}-${gitinfoJson.branch}-${packageJson.version}`;
    return (
        <Card classes={{ root: `${classes.root} card-dashboard-height` }}>
            <CardContent className={classes.version}>
                <Typography variant="body2">version: {version}</Typography>
            </CardContent>
            <CardContent classes={{ root: classes.content }}>
                {loglines &&
                    loglines
                        .slice()
                        .reverse()
                        .map((logline, index) => (
                            <Typography
                                key={index}
                                variant="body2"
                                classes={{ root: classes.message }}
                            >
                                {SeverityIcons[logline.severity]}
                                {logline.message}
                            </Typography>
                        ))}
            </CardContent>
            <CardActions>
                <Button onClick={handleClearLog}>clear</Button>
            </CardActions>
        </Card>
    );
};

export default Log;
