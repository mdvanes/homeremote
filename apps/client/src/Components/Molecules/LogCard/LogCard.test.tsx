import React from "react";
import * as ReactRedux from "react-redux";
import { fireEvent, render } from "@testing-library/react";
import LogCard from "./LogCard";
import { RootState } from "../../../Reducers";
import { Logline, Severity } from "./logSlice";
import * as Store from "../../../store";

const mockUseSelectorWith = ({ lines = [] }: { lines?: Logline[] }): void => {
    jest.spyOn(ReactRedux, "useSelector").mockReset();
    jest.spyOn(ReactRedux, "useSelector").mockImplementation((fn) => {
        const mockRootState: Pick<RootState, "loglines"> = {
            loglines: {
                lines,
                urgentMessage: false,
            },
        };
        return fn(mockRootState);
    });
};

describe("LogCard", () => {
    const mockDispatch = jest.fn();

    beforeEach(() => {
        jest.spyOn(Store, "useAppDispatch").mockReturnValue(mockDispatch);
    });

    it("shows the application version", () => {
        mockUseSelectorWith({});
        const { getByText } = render(<LogCard />);
        expect(getByText(/version:/).textContent).toContain("version: ");
    });

    it("shows info logline with icon", () => {
        mockUseSelectorWith({
            lines: [{ message: "my info message", severity: Severity.INFO }],
        });
        const { getByText } = render(<LogCard />);
        expect(getByText(/my info message/)).toMatchSnapshot();
    });

    it("shows error logline with icon", () => {
        mockUseSelectorWith({
            lines: [{ message: "my error message", severity: Severity.ERROR }],
        });
        const { getByText } = render(<LogCard />);
        expect(getByText(/my error message/)).toMatchSnapshot();
    });

    // TODO fireEvent is not handled after migration to mui5
    it.skip("clears log on button click", () => {
        mockUseSelectorWith({});
        const { getByText } = render(<LogCard />);
        const clearButton = getByText("clear").parentElement;
        mockDispatch.mockReset();
        if (clearButton) {
            fireEvent.click(clearButton);
        }
        expect(mockDispatch).toHaveBeenCalledWith(
            expect.objectContaining({
                type: "loglines/clearLog",
            })
        );
    });
});
