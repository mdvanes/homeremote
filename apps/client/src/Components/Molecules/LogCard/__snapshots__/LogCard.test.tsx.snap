// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogCard shows error logline with icon 1`] = `
<p
  class="MuiTypography-root makeStyles-message MuiTypography-body2"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root"
    focusable="false"
    style="fill: red;"
    viewBox="0 0 24 24"
  >
    <path
      d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
    />
  </svg>
  my error message
</p>
`;

exports[`LogCard shows info logline with icon 1`] = `
<p
  class="MuiTypography-root makeStyles-message MuiTypography-body2"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root"
    focusable="false"
    style="fill: grey;"
    viewBox="0 0 24 24"
  >
    <path
      d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
    />
  </svg>
  my info message
</p>
`;
